File: src/gn/ffi/source_file.cc
// Copyright 2026 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "gn/ffi/source_file.h"

#include "gn/output_file.h"
#include "gn/settings.h"
#include "gn/source_file.h"

rust::Str source_file_to_output_path(const Settings& settings,
                                     const SourceFile& file) {
  OutputFile output_file(settings.build_settings(), file);
  return rust::Str(output_file.value());
}

File: src/gn/starlark/crates/ffi/src/source_file.rs
// Copyright 2026 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use crate::SourceFile;

impl SourceFile {
    /// Converts a GN SourceFile to a starlark File relative to the build
    /// directory.
    pub fn to_rust(&self, settings: &crate::Settings) -> types::File {
        // Safety: source_file_to_output_path returns a string backed by an interned
        // StringAtom and is static.
        types::File::new(unsafe {
            types::util::extend_lifetime(crate::bridge::source_file_to_output_path(settings, self))
        })
    }
}
